cmd/cmd_test.go

```diff
@@ -0,0 +1,315 @@
+			split("config add-context --user user --endpoint endpoint --name my-context")},
+		{"config-delete-context-cascade",
+			split("config delete-context my-context --cascade")},
+		{"config-view",
+			split("config view")},
+		{"config-view",
+			split("config view --raw")},
+
+		{"neg-config-add-user-bad-auth-type",
+			split("config add-user --username foobar --auth-type badbad")},
+	}
+
+	if len(encryptPassword) > 2 { // are we on a platform that supports encryption
+		tests = append(tests, test{"neg-config-add-user-bad-use-encrypted",
+			split(fmt.Sprintf("config add-user --username foobar --auth-type other%v", encryptPassword))})
+	}
+
+	run(t, tests)
+}
+
+func TestGetTags(t *testing.T) {
+	setup(t.Name())
+	tests := []test{
+		{"get-tags",
+			split("install mssql get-tags")},
+	}
+
+	run(t, tests)
+}
+
+func TestMssqlInstall(t *testing.T) {
+	setup(t.Name())
+	tests := []test{
+		{"install",
+			split(fmt.Sprintf("install mssql%v --user-database my-database --accept-eula%v", useCached, encryptPassword))},
+		{"config-current-context",
+			split("config current-context")},
+		{"config-connection-strings",
+			split("config connection-strings")},
+		{"query",
+			split("query GO")},
+		{"query",
+			split("query")},
+		{"neg-query-two-queries",
+			split("query bad --query bad")},
+
+		/* How to get code coverage for user input
+		{"neg-uninstall-no-yes",
+			split("uninstall")},*/
+		{"uninstall",
+			split("uninstall --yes --force")},
+	}
+
+	run(t, tests)
+}
+
+func runTests(t *testing.T, tt struct {
+	name string
+	args struct{ args []string }
+}) {
+	cmd := cmdparser.New[*Root](root.SubCommands()...)
+	cmd.ArgsForUnitTesting(tt.args.args)
+
+	t.Logf("Running: %v", tt.args.args)
+
+	if tt.name == "neg-config-add-user-no-password" {
+		os.Setenv("SQLCMD_PASSWORD", "")
+	} else {
+		os.Setenv("SQLCMD_PASSWORD", "badpass")
+	}
+
+	// If test name starts with 'neg-' expect a Panic
+	if strings.HasPrefix(tt.name, "neg-") {
+		defer func() {
+			if r := recover(); r == nil {
+				t.Errorf("The code did not panic")
+			}
+		}()
+		cmd.Execute()
+	}
+	cmd.Execute()
+}
+
+func Test_displayHints(t *testing.T) {
+	displayHints([]string{"Test Hint"})
+}
+
+func TestIsValidRootCommand(t *testing.T) {
+	Initialize()
+	IsValidSubCommand("install")
+	IsValidSubCommand("create")
+	IsValidSubCommand("nope")
+}
+
+func TestRunCommand(t *testing.T) {
+	loggingLevel = 4
+	Execute()
+}
+
+func Test_checkErr(t *testing.T) {
+	loggingLevel = 3
+
+	defer func() {
+		if r := recover(); r == nil {
+			t.Errorf("The code did not panic")
+		}
+	}()
+
+	checkErr(errors.New("Expected error"))
+}
+
+func run(t *testing.T, tests []test) {
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) { runTests(t, tt) })
+	}
+
+	verifyConfigIsEmpty(t)
+}
+
+func verifyConfigIsEmpty(t *testing.T) {
+	if !config.IsEmpty() {
+		bytes := output.Struct(config.GetRedactedConfig(true))
+		t.Errorf("Config is not empty. Content of config file:\n%s\nConfig file used:%s",
```

Comment:
>t.Errorf [](http://example.com/codeflow?start=2&length=8)

github.com/stretchr/testify/assert is a pretty good assert library

Reply:
Added to:

[https://github.com/microsoft/go-sqlcmd/issues/171](https://github.com/microsoft/go-sqlcmd/issues/171)


Should tests standarize on a specific assert library e.g. github.com/stretchr/testify/assert

Also in the Go Style Guide, what does "invalid local style" / "Use of assertion-based testing libraries" mean:

[https://google.github.io/styleguide/go/guide#local-consistency](https://google.github.io/styleguide/go/guide#local-consistency)

"
Examples of invalid local style considerations:

Line length restrictions for code
Use of assertion-based testing libraries
"